### PR TITLE
Per player viewdistance

### PR DESCRIPTION
**Pull request by Spottedleaf**
See commits

The patch was checked against the original
This patch should be merged after https://github.com/PaperMC/Paper/pull/2957

### CODE REVIEW

**Comment by Aaron2550** [comment]
Quick question: will there be a way to set the maximum regardless?

**Comment by electronicboy** [comment]
This offers API, up to plugins for a "max"

File: Spigot-Server-Patches/0399-implement-optional-per-player-mob-spawns.patch [hunk 1]
```diff
@@ -643,36 +121,67 @@ index 79f854966..4100e367a 100644
...
++
+     void addPlayerToDistanceMaps(EntityPlayer player) {
+         this.updateMaps(player);
+ 
+@@ -116,6 +120,9 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+     }
+ 
+     void removePlayerFromDistanceMaps(EntityPlayer player) {
++        // Paper start - per player mob spawning
++        this.playerMobDistanceMap.remove(player);
```
Comment (Black-Hole) [inline Spigot-Server-Patches/0399-implement-optional-per-player-mob-spawns.patch:152]: this.playerMobDistanceMap might be null here if per player mob spawn is disabled.

Comment (Spottedleaf) [reply Spigot-Server-Patches/0399-implement-optional-per-player-mob-spawns.patch:152]: correct

**Comment by Spottedleaf** [comment]
put on hold until no-tick view distance is implemented

**Comment by rayanbzd** [comment]
Any news ?

**Review by jewome62 — APPROVED** [review]

**Comment by JMCY26** [comment]
> put on hold until no-tick view distance is implemented

Curious by this, is there somewhere we can read more about this implementation to paper?

**Comment by MiniDigger** [comment]
@JMCY26 its implemented here https://github.com/Spottedleaf/Tuinity/blob/master/patches/server/0012-Per-player-view-distance-implementation.patch

**Comment by JMCY26** [comment]
@MiniDigger 
Thank you, loaded it up and it's perfect, long been needing something like this.

**Comment by Spottedleaf** [comment]
busy fixing some of its issues

**Comment by electronicboy** [comment]
Has this been checked for memory leaks? Just asking for a friend

**Comment by metalshark** [comment]
@Spottedleaf has this been abandoned?

**Comment by rayanbzd** [comment]
Available on tuinity fork. 

> @Spottedleaf has this been abandoned?

**Comment by developStorm** [comment]
Why this couldn't be merged into Paper?

**Comment by MiniDigger** [comment]
because it still has issues, it breaks citizens for example iirc

**Comment by A248** [comment]
Just curious, how does it break citizens?

From the Citizens source, they use NMS considerably. If a Paper patch breaks Citizens' NMS, isn't that their own issue for using NMS to begin with? Correct me if their NMS usage is not that which is broken by this PR.

**Comment by MiniDigger** [comment]
citizens uses a custom entity tracker

**Comment by Toffikk** [comment]
@Spottedleaf  Will this patch be ever reimpemented , in that way so it don't break NMS plugins like citizens?

**Pull request closed without merging**